src/main/xerus/monstercat/tabs/TabCatalog.kt

```diff
@@ -55,12 +62,68 @@ class TabCatalog: TableTab() {
 		table.visibleLeafColumns.addListener(ListChangeListener {
 			it.next(); Settings.VISIBLECATALOGCOLUMNS.putMulti(*it.addedSubList.map { it.text }.toTypedArray())
 		})
+		
+		fun playTracks(add: Boolean) {
+			val selected = table.selectionModel.selectedItems ?: return
+			GlobalScope.launch {
+				if(!add)
+					Player.playTracks(getSongs(selected))
+				else
+					Playlist.addAll(getSongs(selected))
+			}
+		}
 		table.setOnMouseClicked { me ->
-			if(me.clickCount == 2) {
-				val selected = table.selectionModel.selectedItem ?: return@setOnMouseClicked
-				Player.play(selected[cols.findUnsafe("Track")].trim(), selected[cols.findUnsafe("Artist")])
+			when {
+				me.clickCount == 2 && me.button == MouseButton.PRIMARY -> playTracks(false)
+				me.button == MouseButton.MIDDLE -> playTracks(true)
+			}
+		}
+		table.setOnKeyPressed { ke ->
+			when(ke.code) {
+				KeyCode.ENTER -> playTracks(false)
+				KeyCode.PLUS, KeyCode.ADD -> playTracks(true)
+				else -> return@setOnKeyPressed
+			}
+		}
+		
+		table.selectionModel.selectionMode = SelectionMode.MULTIPLE
+		
+		val rightClickMenu = ContextMenu()
+		rightClickMenu.items.addAll(
+			MenuItem("Play") {
+				val selected = table.selectionModel.selectedItems
+				GlobalScope.launch {
+					Player.playTracks(getSongs(selected))
+				}
+			},
+			MenuItem("Add to playlist") {
+				val selected = table.selectionModel.selectedItems
+				GlobalScope.launch {
+					Playlist.addAll(getSongs(selected))
+				}
+			},
+			MenuItem("Play next") {
+				val selected = table.selectionModel.selectedItems
+				GlobalScope.launch {
+					Playlist.addAll(getSongs(selected), true)
+				}
+			},
+			SeparatorMenuItem(),
+			MenuItem("Select All") {
+				table.selectionModel.selectAll()
 			}
+		)
+		table.contextMenu = rightClickMenu
+	}
+	
+	private suspend fun getSongs(songList: ObservableList<List<String>>): ArrayList<Track> {
+		val tracklist = arrayListOf<Track>()
+		songList.forEach { item ->
+			val track = APIUtils.find(item[cols.findUnsafe("Track")].trim(), item[cols.findUnsafe("Artist")])
```

Comment:
what about this instead?
```kotlin
APIUtils.find(item[cols.findUnsafe("Track")].trim(), item[cols.findUnsafe("Artist")])?.let {
	tracklist.add(it)
} ?: logger.warn("Failed matching song ${item[cols.findUnsafe("Artist")]} - ${item[cols.findUnsafe("Track")].trim()} while adding it to playlist")
```

Reply:
Harder to read imo

Reply:
4ed611f95efa3ceadf2c54696dd9a6780fe4ea1e